src/index.js

```diff
@@ -16,102 +14,151 @@ function makeDnsPromise(host) {
+    originalEvent,
+    originalContext,
+    originalCallback
+  ) {
+    this.startTime = process.hrtime();
+    this.config = config;
+    this.metrics = [];
+
+    // assign a new dnsPromise if it's not a coldstart because dns could have changed
+    if (!globals.COLDSTART) {
+      this.dnsPromise = getDnsPromise(this.config.host);
+    } else {
+      this.dnsPromise = dnsPromise;
     }
 
-    /* resolve DNS early on coldstarts */
-    dnsPromise = makeDnsPromise(config.host);
+    this.originalContext = originalContext;
+    this.originalCallback = originalCallback;
 
-    return function IOpipeWrapper() {
-      fn.metricsQueue = [];
-      const args = [].slice.call(arguments);
-
-      if (!globals.COLDSTART) {
-        /* Get an updated DNS record. */
-        dnsPromise = makeDnsPromise(config.host);
-      }
-
-      const startTime = process.hrtime();
-      const report = new Report(
-        config,
-        args[1],
-        startTime,
-        fn.metricsQueue,
-        dnsPromise
+    // preserve original functions via a property name change
+    ['succeed', 'fail', 'done'].forEach(method => {
+      Object.defineProperty(
+        this.originalContext,
+        `original_${method}`,
+        Object.getOwnPropertyDescriptor(this.originalContext, method)
       );
+    });
 
-      const timeout = setupTimeoutCapture(config, report, args[1]);
-
-      const callback = (err, cb) => {
-        if (timeout) {
-          clearTimeout(timeout);
-        }
-        report.send(err, cb);
-      };
-
-      /* Mangle arguments, wrapping callbacks. */
-      args[1] = Context(callback, args[1]);
-      args[2] = Callback(callback, args[2]);
-
-      try {
-        return func.apply(this, args);
-      } catch (err) {
-        clearTimeout(timeout);
-        report.send(err, function noop() {});
-        return undefined;
+    // assign modified methods and objects here
+    this.modifiedContext = Object.assign(this.originalContext, {
+      // need to use .bind, otherwise, the this ref inside of each fn is NOT IOpipeWrapperClass
+      succeed: this.succeed.bind(this),
+      fail: this.fail.bind(this),
+      done: this.done.bind(this),
+      iopipe: {
+        log: this.log.bind(this),
+        metrics: this.metrics,
+        version: globals.VERSION,
+        config: this.config
       }
+    });
+
+    this.modifiedCallback = (err, data) => {
+      this.sendReport(err, () => {
+        typeof this.originalCallback === 'function' &&
+          this.originalCallback(err, data);
+      });
     };
-  };
 
-  // Alias decorate to the wrapper function
-  fn.decorate = fn;
+    this.timeout = setupTimeoutCapture(this);
 
-  fn.log = (name, value) => {
+    this.report = new Report(this);
+
+    try {
+      return userFunc.call(
+        this,
+        originalEvent,
+        this.modifiedContext,
+        this.modifiedCallback
+      );
+    } catch (err) {
+      this.sendReport(err);
+    }
+  }
+  sendReport(err, cb = () => {}) {
```

Comment:
not a "no" but I like how we applied DRY here and it is now lost with this PR. Not a big issue though.